clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3091,6 +3104,89 @@ static void handleMaxGlobalWorkDimAttr(Sema &S, Decl *D,
                                                                      E);
 }
 
+SYCLIntelLoopFuseAttr *
+Sema::mergeSYCLIntelLoopFuseAttr(Decl *D, const AttributeCommonInfo &CI,
+                                 Expr *E) {
+
+  if (const auto ExistingAttr = D->getAttr<SYCLIntelLoopFuseAttr>()) {
+    // [[intel::loop_fuse]] and [[intel::loop_fuse_independent]] are
+    // incompatible
+    if (ExistingAttr->getAttributeSpellingListIndex() !=
+        CI.getAttributeSpellingListIndex()) {
+      Diag(CI.getLoc(), diag::err_attributes_are_not_compatible)
+          << CI.getAttrName() << ExistingAttr;
+      Diag(ExistingAttr->getLocation(), diag::note_conflicting_attribute);
+      return nullptr;
+    }
+
+    if (!E->isValueDependent()) {
+      Optional<llvm::APSInt> ArgVal = E->getIntegerConstantExpr(Context);
+      Optional<llvm::APSInt> ExistingArgVal =
+          ExistingAttr->getValue()->getIntegerConstantExpr(Context);
+
+      // Compare attribute argument value and warn if there is a mismatch
+      if (ArgVal->getExtValue() != ExistingArgVal->getExtValue())
+        Diag(ExistingAttr->getLoc(), diag::warn_duplicate_attribute)
+            << ExistingAttr;
+    }
+
+    // If there is no mismatch, silently ignore duplicate attribute
+    return nullptr;
+  }
+  return ::new (Context) SYCLIntelLoopFuseAttr(Context, CI, E);
+}
+
+static bool checkSYCLIntelLoopFuseArgument(Sema &S,
+                                           const AttributeCommonInfo &CI,
+                                           Expr *E) {
+  // Dependent expressions are checked when instantiated.
+  if (E->isValueDependent())
+    return false;
+
+  Optional<llvm::APSInt> ArgVal = E->getIntegerConstantExpr(S.Context);
+  if (!ArgVal) {
+    S.Diag(E->getExprLoc(), diag::err_attribute_argument_type)
+        << CI.getAttrName() << AANT_ArgumentIntegerConstant
+        << E->getSourceRange();
+    return true;
+  }
+
+  SYCLIntelLoopFuseAttr TmpAttr(S.Context, CI, E);
+  ExprResult ICE;
+  if (S.checkRangedIntegralArgument<SYCLIntelLoopFuseAttr>(E, &TmpAttr, ICE))
+    return true;
+
+  return false;
+}
+
+void Sema::addSYCLIntelLoopFuseAttr(Decl *D, const AttributeCommonInfo &CI,
+                                    Expr *E) {
+  assert(E && "argument has unexpected null value");
+
+  if (checkSYCLIntelLoopFuseArgument(*this, CI, E))
+    return;
+
+  SYCLIntelLoopFuseAttr *NewAttr = mergeSYCLIntelLoopFuseAttr(D, CI, E);
+
+  if (NewAttr)
+    D->addAttr(NewAttr);
+}
+
+// Handles [[intel::loop_fuse]] and [[intel::loop_fuse_independent]].
+static void handleLoopFuseAttr(Sema &S, Decl *D, const ParsedAttr &Attr) {
+  // Attribute should not be processed during host compilation.
+  if (S.LangOpts.SYCLIsHost)
```

Comment:
We don't even want to check the validity of the attribute in this mode? I would have imagined that we'd still want to do the semantic checks even if we don't add the attribute, but I am not super familiar with this compilation model.

Reply:
There are I think 4 (or more?)  compilation passes - 
1. Generate .bc file (compiling for device)
2. Generate integration header
3. Create host object using integration header
4. Create bundled object  using 1. and 3. 

I can take another look to see how other attributes are handled but if we ignore a particular attribute during host compilation pass, is there a reason (that I'm unaware of) why we would need to test it's validity?

Reply:
If the user can execute the pass directly then it means they could write something like `[[intel::loop_fuse(1.3f)]]` and the compilation pass would be silent about the error. I have no idea if this is a real problem we need to worry about though, so I'd ignore my comment for now. We can always address it for all the attributes at once if we think a change needs to happen.

Reply:
The user can execute host pass using -fsycl-is-host. (I added a test for this in PR - loop_fusion_host.cpp). This however sets the SYCLIsHost LangOpt, and the attribute is ignored.